### PR TITLE
`heap-to-stack` pass

### PR DESCRIPTION
**Pull request by adam-smnk**
Adds a new pass `heap-to-stack` which converts heap allocations to stack.
The pass is added to the default pipeline at the postprocessing stage.

### CODE REVIEW

File: lib/TPP/HeapToStack.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//===- HeapToStack.cpp -------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Tpp/TppUtils.h"
```
Comment (chelini) [inline lib/TPP/HeapToStack.cpp:9]: why is needed? copy pasta?

File: lib/TPP/HeapToStack.cpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Matchers.h"
```
Comment (chelini) [inline lib/TPP/HeapToStack.cpp:17]: same

File: lib/TPP/HeapToStack.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Dialect/Tpp/TppUtils.h"
+#include "TPP/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
```
Comment (chelini) [inline lib/TPP/HeapToStack.cpp:14]: same.

File: lib/TPP/HeapToStack.cpp (hunk 1 above)
Comment (adam-smnk) [reply lib/TPP/HeapToStack.cpp:9]: It's always pasta 😆

File: lib/TPP/HeapToStack.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    unsigned bitwidth = mlir::DataLayout::closest(alloc).getTypeSizeInBits(
+        type.getElementType());
+    unsigned size = type.getNumElements() * bitwidth;
+    if (size > (maxAllocSizeInBytes * 8))
+      return rewriter.notifyMatchFailure(
+          alloc, "Buffer exceeds maximum convertion size");
+
+    // Find matching deallocation operation.
+    // If the lifetime of the given allocation extends beyond the current scope,
+    // the buffer cannot be converted to stack.
```
[the remark this answers is not in the document]
Comment (adam-smnk) [reply lib/TPP/HeapToStack.cpp:54]: Maybe some case that you allocate a buffer and return it on success but deallocate on failure? Not sure if MLIR IR has a way to represent such optional function return or sth.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/HeapToStack.cpp:48]: Ok, I got confused reading the comment.

[the remark this answers is not in the document]
Comment (chelini) [reply lib/TPP/HeapToStack.cpp:48]: (the comment mentions `scope` but I don't see any check-related to this)

**Comment by rengolin** [comment]
Did you check this against MLP examples? Did it replace the stray `alloc`s? Can we have one test that shows it?

**Comment by adam-smnk** [comment]
> Did you check this against MLP examples? Did it replace the stray `alloc`s? Can we have one test that shows it?

Not explicitly in MLP but some of the benchmarks' performance is higher.
I can add a test case with MLP.

File: lib/TPP/HeapToStack.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rengolin) [reply lib/TPP/HeapToStack.cpp]: I wouldn't do a parent region check as this only complicates things and we don't have a use-case for that. We just soft-fail (leave the alloc) in those cases for now.

**Review by rengolin — APPROVED** [review]
Looks simple enough, thanks!

**Pull request merged**